baseapp/test_helpers.go

```diff
@@ -25,6 +25,9 @@ func (app *BaseApp) SimCheck(txEncoder sdk.TxEncoder, tx sdk.Tx) (sdk.GasInfo, *
 
 // Simulate executes a tx in simulate mode to get result and gas info.
 func (app *BaseApp) Simulate(txBytes []byte) (sdk.GasInfo, *sdk.Result, error) {
+	app.simulateMutex.RLock()
```

Comment:
This would block CheckTX 💣

Reply:
yea CheckTx and Simulate have to be mutually locked because CheckTx ante can update `checkState` which can affect Simulate result.

Reply:
Simulate shouldn't need a mutex given that in runTx we do `cacheTxContext` and only write back to disk if we are on finalizeBlock (`msCache.Write()`)

Reply:
I think you need to see the discussion and description on the issue